pyicloud/services/photos.py

```diff
@@ -173,40 +175,67 @@ def albums(self):
-                # TODO: Handle subfolders  # pylint: disable=fixme
                 if folder["recordName"] == "----Root-Folder----" or (
                     folder["fields"].get("isDeleted")
                     and folder["fields"]["isDeleted"]["value"]
                 ):
                     continue
 
-                folder_id = folder["recordName"]
-                folder_obj_type = (
-                    "CPLContainerRelationNotDeletedByAssetDate:%s" % folder_id
-                )
-                folder_name = base64.b64decode(
-                    folder["fields"]["albumNameEnc"]["value"]
-                ).decode("utf-8")
-                query_filter = [
-                    {
-                        "fieldName": "parentId",
-                        "comparator": "EQUALS",
-                        "fieldValue": {"type": "STRING", "value": folder_id},
-                    }
-                ]
-
-                album = PhotoAlbum(
-                    self,
-                    folder_name,
-                    "CPLContainerRelationLiveByAssetDate",
-                    folder_obj_type,
-                    "ASCENDING",
-                    query_filter,
-                )
-                self._albums[folder_name] = album
+                if folder["fields"]["albumType"]["value"] == self.ALBUM_TYPES["album"]:
+                    self._construct_album([], folder)
+                elif (
+                    folder["fields"]["albumType"]["value"] == self.ALBUM_TYPES["folder"]
+                ):
+                    self._construct_folder([folder])
 
         return self._albums
 
+    def _construct_album(self, parents, folder):
+        folder_id = folder["recordName"]
+        folder_obj_type = "CPLContainerRelationNotDeletedByAssetDate:%s" % folder_id
+        folder_name = base64.b64decode(
+            folder["fields"]["albumNameEnc"]["value"]
+        ).decode("utf-8")
+        query_filter = [
+            {
+                "fieldName": "parentId",
+                "comparator": "EQUALS",
+                "fieldValue": {"type": "STRING", "value": folder_id},
+            }
+        ]
+
+        parents_names = [
+            base64.b64decode(p["fields"]["albumNameEnc"]["value"]).decode("utf-8")
+            for p in parents
+        ]
+
+        album = PhotoAlbum(
+            self,
+            folder_name,
+            "CPLContainerRelationLiveByAssetDate",
+            folder_obj_type,
+            "ASCENDING",
+            query_filter,
+            parents_names,
+        )
+        self._albums[folder_name] = album
+
+    def _construct_folder(self, folder_hierarchy):
+        folder = folder_hierarchy[-1]
+        folder_id = folder["recordName"]
+
+        for sub_folder in self._fetch_sub_folders(folder_id):
+            if sub_folder["recordName"] == "----Root-Folder----" or (
+                sub_folder["fields"].get("isDeleted")
+                and sub_folder["fields"]["isDeleted"]["value"]
```

Comment:
Why checking the value here ?

Should we fetch sub folders if the parent is deleted ?

Reply:
Great point; this logic exists already at https://github.com/picklepete/pyicloud/pull/297/files/2e4d3eb62ffc3b36bf3d30e60ac526f48cd366c4#diff-dbc5e8af14fd471636e609e5ecd1fe3bR178. I'll take this out and see what happens.

Reply:
I'm inclined to leave this logic in, in other words, I'm included to not fetch sub folders / albums for deleted parent folders.

In the iCloud GUI, if you delete a folder that contains sub folders, it will delete both the folder and the sub-folder. I think it may be confusing/inconsistent if some deleted folders are traversed when looking for albums, as the results generated by pyicloud will be different than what is visible in the UI. This also remains aligned with the existing code where deleted albums were not returned.

I could see it being valuable to allow the user to choose how deleted folders/abums should be treated when fetching albums in a future PR that extends this functionality.

Reply:
OK but I think we should add a quick comment to explain how it is.

Reply:
Also we can extract the code here in a little private function as, as you pointed out, the logic exists already at L178

Reply:
@Quentame I've implemented both of your suggestions, thank you for making them. Sorry for the slow delay in implementing them!

I think all of the feedback has been addressed and this can be merged. Please let me know if you have other feedback.